networked_controller.cpp

```diff
@@ -1530,211 +1433,140 @@ DollController::DollController(NetworkedController *p_node) :
-	buffer_start_position += 1;
-
-	while (buffer_start_position < p_data.size()) {
-		const int buffer_size = p_data[buffer_start_position];
-		ERR_FAIL_COND_MSG(buffer_size <= 0, "The `buffer_size` is never networked when less than 0, this is a bug.");
-		const Vector<uint8_t> buffer = p_data.slice(
-				buffer_start_position + 1,
-				buffer_start_position + 1 + buffer_size);
-
-#ifdef DEBUG_ENABLED
-		CRASH_COND_MSG(buffer.size() != buffer_size, "The buffer.size() is expected to be the same at this point.");
-#endif
+	if (epoch <= future_epoch) {
+		// This epoch is old, it arrived too late; nothing to do.
+		return;
+	}
 
-		const uint32_t epoch = receive_epoch(buffer);
-		buffer_start_position += 1 + buffer_size;
+	const int64_t current_virtual_delay = int64_t(future_epoch) - int64_t(current_epoch);
 
-		batch_young_epoch = MIN(epoch, batch_young_epoch);
+	if (current_epoch > future_epoch) {
+		// Make sure we set back the current epoch in case of overshot.
+		// The overshot is wanted to correctly calculate `current_virtual_delay`, but at this point
+		// the overshot need to be normalized.
+		current_epoch = future_epoch;
 	}
+	past_epoch = current_epoch;
+	future_epoch = epoch;
 
-	// ~~ Establish the interpolation speed ~~
-	if (batch_young_epoch == UINT32_MAX) {
-		// This may just be a late arrived batch, so nothing more to do.
-		return;
-	}
+	// Make sure to store the current state, so we can use it to interpolate.
+	// This is necessary so we allow a bit of overshot.
+	past_epoch_buffer.begin_write(0);
+	node->native_collect_epoch_data(past_epoch_buffer);
 
-	const real_t net_sentitivity = node->get_doll_net_sensitivity();
+	// ~~ Establish the interpolation speed ~~
 
 	// Establish the connection quality by checking if the batch takes
 	// always the same time to arrive.
 	const uint32_t now = OS::get_singleton()->get_ticks_msec();
+
 	// If now is bigger, then the timer has been disabled, so we assume 0.
-	network_watcher.push(now > batch_receiver_timer ? now - batch_receiver_timer : 0);
-	batch_receiver_timer = now;
 
-	const uint32_t avg_receive_time = network_watcher.average();
-	const real_t deviation_receive_time = real_t(network_watcher.get_deviation(avg_receive_time)) / 1000.0;
+	real_t packet_arrived_in = 0.0;
+	if (now > epoch_received_timestamp) {
+		packet_arrived_in = static_cast<real_t>(now - epoch_received_timestamp) / 1000.0;
+		const real_t delay = packet_arrived_in - next_epoch_expected_in;
+		network_watcher.push(ABS(delay));
+	}
+	epoch_received_timestamp = now;
+	next_epoch_expected_in = next_sync_time;
 
+	const real_t avg_delay = network_watcher.average();
 	// The network quality can be established just by checking the standard
 	// deviation. Stable connections have standard deviation that tend to 0.
-	const real_t net_poorness = MIN(
-			deviation_receive_time / net_sentitivity,
-			1.0);
+	//const real_t deviation_delay = network_watcher.get_deviation(avg_delay);
```

Comment:
I've corrected the variable naming, now it's more clear what's doing: Essentially inside the `network_watcher`, you can find the **difference** between the arrival time and the **expected** time.

If the average is not 0 it means that the connection is oscillating, and that's the reason why I just used the average. However, now I changed it (and re-balanced the settings) so that it also uses the deviation. In this way the check is more accurate.